### PR TITLE
bpo-44035: Check autoconf files thoroughly (GH-29935)

### PR DESCRIPTION
**Pull request by tiran**
Check that users don't push changes with outdated or patched autoconf.
The presence of runstatedir option and aclocal 1.16.3 are good markers.

Use my container image to regenerate autoconf files. "Check for changes"
will fail later when any file is regenerated.

Use ccache in check_generated_files to speed up testing.

https://bugs.python.org/issue44035

### CODE REVIEW

**Comment by arhadthedev** [comment]
Can we make one step further and use GitHub API to automatically add a file-wide review with a `suggestion` code block if after a checkout of a tracking branch and the regeneration into its working dir, `git diff` returns something?

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -64,6 +64,20 @@ jobs:
       - uses: actions/setup-python@v2
       - name: Install Dependencies
         run: sudo ./.github/workflows/posix-deps-apt.sh
+      - name: Add ccache to PATH
+        run: |
+          echo "PATH=/usr/lib/ccache:$PATH" >> $GITHUB_ENV
```
Comment (erlend-aasland) [inline .github/workflows/build.yml]:
```suggestion
        run: export PATH="/usr/lib/ccache:$PATH"
```

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -64,6 +64,20 @@ jobs:
...
+      - name: Configure ccache action
+        uses: hendrikmuhs/ccache-action@v1
+      - name: Check Autoconf version 2.69 and aclocal 1.16.3
+        run: |
+          set -e
```
Comment (erlend-aasland) [inline .github/workflows/build.yml]: GitHub already runs `bash` with `-e`.

```suggestion
```

**Comment by tiran** [comment]
> Can we make one step further and use GitHub API to automatically add a file-wide review with a `suggestion` code block if after a checkout of a tracking branch and the regeneration into its working dir, `git diff` returns something?

That's an interesting idea and certainly helpful for new contributors. Could you please open a new ticket for your proposal?

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -64,6 +64,20 @@ jobs:
...
+          grep "Generated by GNU Autoconf 2.69" configure
+          grep "runstatedir" configure
+          grep "PKG_PROG_PKG_CONFIG" aclocal.m4
```
Comment (erlend-aasland) [inline .github/workflows/build.yml]: Personal preference; less noise in the output. The other two `grep`s are fine as they are, IMO.

```suggestion
          grep -q "runstatedir" configure
          grep -q "PKG_PROG_PKG_CONFIG" aclocal.m4
```

**Comment by erlend-aasland** [comment]
Why not just add a `regen-autoconf` Make target, add that to `make regen-all`, and let the _Check for changes_ step work it's magic?

Something like this:
```diff
diff --git a/Makefile.pre.in b/Makefile.pre.in
index 8e6e553554..5afe48a26f 100644
--- a/Makefile.pre.in
+++ b/Makefile.pre.in
@@ -1156,6 +1156,10 @@ Python/frozen_modules/getpath.h: $(FREEZE_MODULE) Modules/getpath.py
 
 Tools/scripts/freeze_modules.py: $(FREEZE_MODULE)
 
+.PHONY: regen-autoconf
+regen-autoconf: configure.ac configure aclocal.m4
+       @docker run --rm -v $(shell pwd):/src quay.io/tiran/cpython_autoconf:269
+
 .PHONY: regen-frozen
 regen-frozen: Tools/scripts/freeze_modules.py $(FROZEN_FILES_IN)
        $(PYTHON_FOR_REGEN) $(srcdir)/Tools/scripts/freeze_modules.py
@@ -1176,7 +1180,7 @@ regen-limited-abi: all
 
 regen-all: regen-opcode regen-opcode-targets regen-typeslots \
        regen-token regen-ast regen-keyword regen-frozen clinic \
-       regen-pegen-metaparser regen-pegen regen-test-frozenmain
+       regen-pegen-metaparser regen-pegen regen-test-frozenmain regen-autoconf
        @echo
        @echo "Note: make regen-stdlib-module-names and autoconf should be run manually"
 

```

**Comment by tiran** [comment]
Because it's not that simple. Platforms with mandatory access control and LSMs need extra options. For example SELinux enforcing systems require ``:Z``. There are also distros that don't ship docker. My distro of choice does not have a docker command unless I install the optional ``podman-docker`` shim package.

**Comment by erlend-aasland** [comment]
> Because it's not that simple. Platforms with mandatory access control and LSMs need extra options. For example SELinux enforcing systems require `:Z`. There are also distros that don't ship docker. My distro of choice does not have a docker command unless I install the optional `podman-docker` shim package.

Ah, true. I was only thinking of the CI, but of course the Makefile has a wider audience.

**Review by erlend-aasland — APPROVED** [review]
LGTM!

**Review by pablogsal — APPROVED** [review]
LGTM

**Pull request merged**